affectedmoduledetector/src/main/kotlin/com/dropbox/affectedmoduledetector/ToStringLogger.kt

```diff
@@ -20,52 +20,95 @@
 ) : OutputEventListenerBackedLogger(
         "amd",
-        OutputEventListenerBackedLoggerContext(
-                Clock {
-                    System.currentTimeMillis()
-                }
-        ).also {
+        OutputEventListenerBackedLoggerContext {
+            System.currentTimeMillis()
+        }.also {
             it.level = LogLevel.DEBUG
-            it.setOutputEventListener {
-                stringBuilder.appendln(it.toString())
+            it.setOutputEventListener { outputEvent ->
+                stringBuilder?.get()?.parameters?.getStringBuilderProperty()?.get()?.appendLine(outputEvent.toString())
             }
         },
         Clock {
             System.currentTimeMillis()
         }
 ) {
+
     /**
      * Returns the current log.
      */
-    fun buildString() = stringBuilder.toString()
+    fun buildString() = stringBuilder?.get()?.parameters?.getStringBuilderProperty()?.get()?.toString()
 
+    @Suppress("UnstableApiUsage") // BuildService is not yet stable
     companion object {
+        internal abstract class ToStringLoggerBuildService : BuildService<ToStringLoggerBuildService.ToStringLoggerBuildServiceParameters>, BuildOperationListener, AutoCloseable {
+            interface ToStringLoggerBuildServiceParameters : BuildServiceParameters {
+                fun getStringBuilderProperty(): Property<StringBuilder>
+                fun getOutputFileProperty(): RegularFileProperty
+            }
+
+            override fun started(p0: BuildOperationDescriptor, p1: OperationStartEvent) { }
+
+            override fun progress(p0: OperationIdentifier, p1: OperationProgressEvent) { }
+
+            override fun finished(buildOperationDescriptor: BuildOperationDescriptor, operationFinishEvent: OperationFinishEvent) { }
+
+            override fun close() {
+                val outputFile = parameters.getOutputFileProperty().orNull?.asFile ?: return
+                outputFile.appendText(parameters.getStringBuilderProperty().get().toString())
+                println("Wrote dependency log to ${outputFile.absolutePath}")
+            }
+        }
+
+        /**
+         * Creates the [ToStringLogger]
+         *
+         * @param project the current project to apply to
+         * @param logFilename the filename for the logs to go
+         * @param logFolder the path to where the log should output. if null doesn't output
+         */
         fun createWithLifecycle(
-            gradle: Gradle,
-            onComplete: (String) -> Unit
+            project: Project,
+            logFilename: String,
+            logFolder: String? = null
         ): Logger {
-            val logger = ToStringLogger()
-            gradle.addBuildListener(object : BuildAdapter() {
-                override fun buildFinished(result: BuildResult) {
-                    onComplete(logger.buildString())
+            val gradle = project.gradle
+            val stringBuilder = StringBuilder()
```

Comment:
for wat we are using stringBuilder? We doesn't use result of one. Hence, I gess we can delete it.